### ISSUE TITLE
`env_prefix` falls back to env vars without prefix

### ISSUE THREAD

**Issue by hofrob** [pydantic/pydantic-settings#441]
Using my example from #437: https://github.com/hofrob/pydantic-settings-env-prefix

If you `uv run hello` the result is:

```
hello='ABC'
```

This value is coming from the first line in `.env`: `HELLO=ABC`

I don't think the intention of `env_prefix` is to just fall back to whatever env var it can find that matches the variable without the prefix. If you have nested or complex settings with multiple API keys managed via different `env_prefix` values, you'll not want to use an API key with the wrong service. I'd consider this API key compromised.

**Comment by ericchansen** [comment]
I'm confused about env_prefix. When I use env_prefix, the original env var gets ignored. It will only load an env var if it it matches the name with the env_prefix.

**Comment by hofrob** [comment]
Yes, it works when there an env var exists with the prefix. But I think it should stay empty if it does not find the env var + prefix as opposed to fall back to an env var without the prefix.

- `SOME_PREFIX_HELLO=FOO uv run hello`: `hello='FOO'` :heavy_check_mark: 
- `HELLO=FOO uv run hello`: `hello='ABC'` :x: 
- `uv run hello`: `hello='ABC'` :x:

**Comment by ericchansen** [comment]
Thanks! I thought that setting env_prefix acted like an alias, such that in your example, either `HELLO` or `SOME_PREFIX_HELLO` should work. I guess the documentation is confusing me. It would be nice to get a clear answer from the devs as to what the intended behavior is.

I'm currently struggling with pydantic as well, but it seems like we're observing different behavior. In my use case, I was expecting the env var to get loaded from the raw variable or the variable name with the env_prefix. However, it ONLY works for the variable name with the env_prefix (the raw variable name doesn't work). @hofrob, does this sounds like the opposite problem that you're having?

**Comment by apowers313** [comment]
It seems like the intuition people have about .env is that gets treated the exact same way as the OS environment variables. Maybe it would make sense to merge both `os.environ` and `python-dotenv` into a single dict and then pass that into the config, rather than treating them differently?

**Comment by hofrob** [comment]
I'm not sure I follow. Are the prefixes treated differently depending on the source?

The main point about this issue is that `pydantic-settings` should not fall back in case there is no env var with the defined prefix.

**Comment by emilhem** [comment]
Here's a fork with a test capturing this fault (no fix though): https://github.com/emilhem/pydantic-settings/tree/dotenv_prefix_fix_list

I believe a check is needed here to check if the variable `env_value` is written to a `env_name` which has to be prefixed.
https://github.com/pydantic/pydantic-settings/blob/65929cd1f5356b967456ac1b4f78e929fe04d398/pydantic_settings/sources.py#L1060

**Comment by tselepakis** [comment]
Hi everyone,

I've opened PR #648 to address this issue.

The idea is simple: if an env_prefix is set and we find an environment variable whose name matches a model field but lacks the prefix, we skip it. This avoids unexpected overrides and ensures prefixed variables are prioritized.

I'd love to hear your feedback on the approach or any edge cases you'd like to see covered.

Thanks!

**Comment by hofrob** [comment]
I can only tell from the tests that this looks ok to me!

**Comment by hramezani** [comment]
The fix has been released in `pydantic-settings 2.11.0`. We might need to revert it if it breaks user's existing code.
